### PR TITLE
#63 add context rule

### PR DESCRIPTION
**Pull request by abynenkov-ib**

### CODE REVIEW

**Comment by rchowinfoblox** [comment]
ABAC-policy requires this context rule stanza:
```
context {
    where ctx.tenant == "acme.com";
} {
    allow subject user pete.rose@chicago.il.us to buy products.inventory where ctx.genre == "gambling";
}
```

File: pkg/ast/ast.go [hunk 1]
```diff
@@ -120,6 +120,28 @@ func (a *ActionStatement) String() string {
...
+	Where   *WhereClause
+}
+type ContextAction struct {
+	Action      *Identifier
+	TypePattern *Identifier
+}
+type ContextStatement struct {
+	Token token.Token
+
+	Contidions []*ContextCondition
```
Comment (wplu) [inline pkg/ast/ast.go:134]:
```suggestion
	Conditions []*ContextCondition
```

File: pkg/ast/ast.go [hunk 2]
```diff
@@ -120,6 +120,34 @@ func (a *ActionStatement) String() string {
...
+	Action      *Identifier
+	Subject     Subject
+	Verb        *Identifier
+	TypePattern *Identifier
+	Where       *WhereClause
+}
+type ContextStatement struct {
+	Token token.Token
+
+	Contidions  []*ContextCondition
```
Comment (wplu) [inline pkg/ast/ast.go:139]:
```suggestion
	Conditions  []*ContextCondition
```

File: pkg/compiler/rego/rego.go [hunk 3]
```diff
@@ -43,11 +43,19 @@ func (c *CompilerRego) Compile(pkgname string, pols *ast.Policies) (string, erro
...
 
 	var lineNum int
+	var out string
+	var err error
 	for idx, stmt := range pols.Statements {
 		lineNum += 1
 		switch stmt.(type) {
 		case *ast.ActionStatement:
-			out, err := c.compileStatement(stmt.(*ast.ActionStatement), lineNum)
+			out, err = c.compileStatement(stmt.(*ast.ActionStatement), lineNum)
```
Comment (wplu) [inline pkg/compiler/rego/rego.go:52]:
```suggestion
			out, err := c.compileStatement(stmt.(*ast.ActionStatement), lineNum)
```

File: pkg/compiler/rego/rego.go [hunk 4]
```diff
@@ -124,6 +132,79 @@ func (c *CompilerRego) compileSetDefaults(val string, ids ...string) []string {
...
+					ctx.Contidions = append(ctx.Contidions, icond)
+				}
+				line = append(line, c.linearizeContext(ctx)...)
+			}
+		}
+	}
+	return line
+}
+
+func (c *CompilerRego) compileContextStatement(stmt *ast.ContextStatement, lineNum int) (string, error) {
```
Comment (wplu) [inline pkg/compiler/rego/rego.go:189]: TODO: future pr - lineNum may need sublineNum also, 
statement 1 is broken up into 1a, 1b, 1c, ...

File: pkg/compiler/rego/rego.go (hunk 3 above)
Comment (abynenkov-ib) [reply pkg/compiler/rego/rego.go]: I have another suggestion. Is it looks better now?

File: pkg/compiler/rego/rego.go [hunk 5]
```diff
@@ -124,8 +130,82 @@ func (c *CompilerRego) compileSetDefaults(val string, ids ...string) []string {
...
+
+		rego += cs + "\n"
+	}
+
+	return rego, err
+}
+
 // compileStatement converts the AST statement to a string
-func (c *CompilerRego) compileStatement(stmt *ast.ActionStatement, lineNum int) (string, error) {
+func (c *CompilerRego) compileStatement(stmt *ast.ActionStatement, lineNum *int) (string, error) {
```
Comment (wplu) [inline pkg/compiler/rego/rego.go:208]: let's not use pointer lineNum here and to callees

File: pkg/compiler/rego/rego.go [hunk 6]
```diff
@@ -206,7 +286,7 @@ func (c *CompilerRego) String() string {
 	return fmt.Sprintf("compiler for %s language", Language)
 }
 
-func (c *CompilerRego) compileWhereClause(cnds ast.Condition, lineNum int) (string, error) {
+func (c *CompilerRego) compileWhereClause(cnds ast.Condition, lineNum *int) (string, error) {
```
Comment (wplu) [inline pkg/compiler/rego/rego.go]:
```suggestion
func (c *CompilerRego) compileWhereClause(cnds ast.Condition, lineNum int) (string, error) {
```

File: pkg/compiler/rego/rego.go (hunk 3 above)
Comment (wplu) [reply pkg/compiler/rego/rego.go:53]: looks great

**Review by wplu — APPROVED** [review]

**Pull request merged**